### PR TITLE
KAFKA-7895: fix stream-time reckoning for Suppress (2.1) (#6286)

### PR DESCRIPTION
**Pull request by vvcephei**
Even within a Task, different Processors have different perceptions
of time, due to record caching on stores and in suppression itself,
and in general, due to any processor logic that may hold onto
records arbitrarily and emit them later. Thanks to this, we can't rely
on the whole task existing in the same "instant" of stream-time. The
solution is for each processor node that cares about stream-time to
track it independently.

On the side, backporting some internally-facing code maintainability updates

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java:44]: This was an unrelated change in trunk/2.2 to improve the code legibility.

I'm proposing to backport it to keep the Suppress code consistent and improve maintainability.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/SuppressedInternal.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/SuppressedInternal.java:47]: Likewise, this is part of the Suppress code legibility change.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @bbejeck Here's the backport to 2.1 of (#6286), when you have a chance to review.

I haven't had a chance to run the tests yet, which we should definitely do before merging, but I just wanted to push my branch and open the PR, since the cherry-pick was so messy.

**Review by bbejeck — APPROVED** [review]
Thanks, @vvcephei for the separate PR for `2.1`.
LGTM, pending passing tests.  

But I do have one question - `AbstractSegments.java` and `KTableSuppressProcessorMetricsTest.java` show up with no changes but in #6286 they had some minor changes.  Is this intentional?

**Comment by guozhangwang** [comment]
> But I do have one question - `AbstractSegments.java` and `KTableSuppressProcessorMetricsTest.java` show up with no changes but in #6286 they had some minor changes. Is this intentional?

`AbstractSegments` in 2.1 does not exist yet (it was refactored in 2.2 only), and `KTableSuppressProcessorMetricsTest` is also only added in 2.2. as well.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
@vvcephei failed with

```
Execution failed for task ':streams:checkstyleMain'.
```

**Comment by vvcephei** [comment]
Thanks for the reviews, @guozhangwang and @bbejeck .

Did that answer your question, @bbejeck ? The reason `AbstractSegments` showed up in the diff was actually a mistake on my part. I pushed an empty file. I've removed it. In the 2.1 branch, all that functionality is part of `Segments`.

Thanks for the catch @guozhangwang . That was due to cherry-pick noise. I've pushed a fix for it now, and will monitor the test results.

**Comment by bbejeck** [comment]
>Did that answer your question, @bbejeck ? The reason AbstractSegments showed up in the diff was actually a mistake on my part. I pushed an empty file. I've removed it. In the 2.1 branch, all that functionality is part of Segments.

Yep, just waiting for a green build at this point.

**Comment by vvcephei** [comment]
Hi @bbejeck  and @guozhangwang , the tests have passed. Is this ready to merge?

Thanks,
john

**Comment by vvcephei** [comment]
Oh, actually no! I need to run the system tests.

all system tests passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-03-05--001.1551765320--vvcephei--KAFKA-7895-fix-suppress-2.1--dcacd02/report.html
smoke test passed: http://testing.confluent.io/confluent-kafka-branch-builder-system-test-results/?prefix=2019-03-05--001.1551756286--vvcephei--KAFKA-7895-fix-suppress-2.1--dcacd02/

**Comment by vvcephei** [comment]
Ok, the system tests passed as well. I think it's ready to merge now.

**Comment by bbejeck** [comment]
merged #6325 into `2.1`

**Comment by vvcephei** [comment]
Thanks, @bbejeck !

**Pull request merged**
